vignettes/programming.Rmd

```diff
@@ -0,0 +1,113 @@
+  `nest()`/`unnest()`, `separate()`/`extract()`, and `unite()` let you select
+  variables based on position, name, or type (e.g. `1:3`, `starts_with("x")`, or `is.numeric`). Literally, you can use all the same techniques as with
+  `dplyr::select()`.
+
+* **Data masking**: `expand()`, `crossing()` and `nesting()` let you refer to
+  use data variables as if they were variables in the environment (i.e. you 
+  write `my_variable` not `df$myvariable`). 
+  
+We focus on tidy selection here, since it's the most common. You can learn more about data masking in the equivalent vignette in dplyr: <https://dplyr.tidyverse.org/dev/articles/programming.html>. For other considerations when writing tidyr code in packages, please see `vignette("in-packages")`.
+
+We've pointed out that tidyr's tidy evaluation interface is optimized for interactive exploration. The flip side is that this adds some challenges to indirect use, i.e. when you're working inside a `for` loop or a function. This vignette shows you how to overcome those challenges. We'll first go over the basics of tidy selection and data masking, talk about how to use them indirectly, and then show you a number of recipes to solve common problems.
+
+Before we go on, we reveal the version of tidyr we're using and make a small dataset to use in examples.
+
+```{r}
+packageVersion("tidyr")
+
+mini_iris <- as_tibble(iris)[c(1, 2, 51, 52, 101, 102), ]
+mini_iris
+```
+
+## Tidy selection
+
+Underneath all functions that use tidy selection is the [tidyselect](http://tidyselect.r-lib.org/) package. It provides a miniature domain specific language that makes it easy to select columns by name, position, or type. For example:
+
+* `select(df, 1)` selects the first column; 
+  `select(1, last_col())` selects the last column.
+  
+* `select(df, c(a, b, c))` selects columns `a`, `b`, and `c`.
+
+* `select(df, starts_with("a"))` selects all columns whose name starts with "a";
+  `select(df, ends_with("z"))` selects all columns whose name ends with "z".
+  
+* `select(df, is.numeric)` selects all numeric columns.
+
+You can see more details in `?tidyr_tidy_select`.
+
+### Indirection
+
+Tidy selection makes a common task easier at the cost of making a less common task harder. When you want to use tidy select indirectly with the column specification stored in an intermediate variable, you'll need to learn some new tools. There are three main cases where this comes up:
+
+*   When you have the tidy-select specification in a function argument, you 
+    must **embrace** the argument by surrounding it in doubled braces.
+
+    ```{r}
+    nest_egg <- function(df, cols) {
+      nest(df, egg = {{ cols }})
+    }
+    
+    nest_egg(mini_iris, -Species)
+    ```
+
+*   When you have a character vector of variable names, you must use `all_of()` 
+    or `any_of()` depending on whether you want the function to error if a 
+    variable is not found. These functions allow you to write for loops or a
```

Comment:
I don't really see the connection to `for` loops (here or in the intro). It feels like talking about the scenarios of "character vector of names" and "writing your own wrapper function" covers it.

Reply:
It's the other form of indirection we traditionally forgot to talk about in (e.g.) dplyr documentation. But I couldn't come up with an example where it would be helpful in tidyr.